### ISSUE TITLE
Revert DropdownV2 requirements, support both signatures for Dropdowns

### ISSUE THREAD

**Issue by joshblack** [IBM/carbon-components-react#1364]
Related to #1228, we want to revert the requirements of DropdownV2 from requiring a data structure of items to render and bring back the ability to render DropdownItem's inside of DropdownV2.

**Comment by justinpurcell1** [comment]
Getting my bearings here. 

I had initially opened issue #1495 to extend the functionality of Dropdown. However, after seeing the deprecation warnings, I created PR #1505 to allow DropdownV2 to accept a function to map items to components. After getting some feedback, I adjusted it to make it more generic, and it mostly works with the current implementation of DropdownItem. It also will also accept other custom components in addition to the existing behavior with strings and objects. 

@tomkdgun does this fulfill your use cases?
@joshblack is this cohesive with existing patterns within the project?

**Comment by joshblack** [comment]
Thanks so much for taking this work on so far @justinpurcell1!

I think something that might not be clear with this issue title is that it'd be great for us to support these two patterns:

```jsx
<Dropdown items={...} itemToString={...} />

<Dropdown>
  <DropdownItem />
  <DropdownItem />
  <DropdownItem />
</Dropdown>
```

If it's unrealistic to have one component have both implementations, that's totally fair. There was an idea that we could use Downshift to power all of our dropdowns but I think that the flexibility of controlling dropdown content is too important for this specific component (or other dropdown variants).

I think the real question becomes how can we leverage `downshift` while providing it's render prop surface area, or should we just try and build these components to be as easy-to-use as possible and be okay with restricting customizability in dropdown items.

**Comment by ZacharyStair** [comment]
@joshblack is the benefit of the second solution that people can use the literal `DropdownItem` components? or that they can mix and match what components they pass as children to the `DropdownV2`? or is it just for ease of use, and would only accept `DropdownItem`s?

Justin's solution could be modified to accept a signature that looks like the latter example, the Dropdown with DropdownItem children: basically using a getter to select either `this.props.items || React.Children.toArray(this.props.children)` as items, then letting the default itemToString get `item.props.itemText` as appropriate. It isn't the ugliest solution but it makes the code a bit more complex.

The main impediment is that the current `DropdownItem` has some idiosyncrasies that make it hard to use in isolation: it renders an `li` at the top level, it preventsDefault on click events, etc.; It might make sense to make a `DropdownItemV2` for optional use in the `DropdownV2`, but that doesn't seem to accomplish anything that the current DropdownV2 signature can't do. There would basically just be 2 different ways to do the same thing (rendering the default `DropdownItem`s as children or as items), and a 3rd way to use if you want to customize the dropdown items. In Justin's pr there's a storybook example that does what I'm describing:
```
        <DropdownV2
          {...props()}
          items={dropdownItems}
          itemToString={item => (item ? item.itemText : '')}
          itemToElement={DropdownItem /* <-- you can pass react components or functions to `itemToElement` */}
          onChange={action('onChange')}
        />
```

I guess my point is that the second signature isn't necessary to achieve the level of customizability that we want, I can't speak for anybody else who was interested in this

**Comment by tomkdgun** [comment]
> ...
> @tomkdgun does this fulfill your use cases?

Base on the description, it should work for us.

**Comment by joshblack** [comment]
@ZacharyStair yup! I mistakenly assumed with the deprecation notice that the API of downshift would match the needs of our group, but that's clearly not the case 😞 

Agreed that `itemToElement` could work, as well. We could also use the `renderItem` prop paradigm, as well, to match other solutions in the industry. Definitely am not opposed to that if that is the preference here.

**Comment by justinpurcell1** [comment]
@joshblack any update here? We're relying on this for an upcoming feature. I think there are a couple of viable options. First would be to continue on with `itemToElement`, second would be to change the prop name to `renderItem` as with FlatList (and others?), and third would be to abandon this path and render children in the old DropdownItem.

Passing in proper DropdownItems to DropdownV2 doesn't really work due to the control flow of Dropdown vs V2.

**Comment by laramlewis** [comment]
@joshblack - Will this enhancement allow you to render React components for the dropdown items? We have a need to render an optional icon as well as two pieces of justified and differently styled text in our dropdown entries.

<img width="356" alt="image" src="https://user-images.githubusercontent.com/13322995/48654896-9c9a7f00-e9c5-11e8-997c-445e6ed61690.png">

Is there an ETA on when this enhancement/feature will be available?

**Comment by tomkdgun** [comment]
@joshblack I see sth was merged. Do issue if fixed ? On which version feature is available ?
